src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs

```diff
@@ -103,75 +105,126 @@ private static bool IsCodeActionNotSupportedByLSP(IUnifiedSuggestedAction sugges
-                    Data = new CodeActionResolveData(currentTitle, codeAction.CustomTags, request.Range, request.TextDocument, fixAllFlavors: null)
-                });
-
-                if (suggestedAction is UnifiedCodeFixSuggestedAction unifiedCodeFixSuggestedAction && unifiedCodeFixSuggestedAction.FixAllFlavors is not null)
+                if (!isTopLevelCodeAction)
                 {
-                    var fixAllFlavors = unifiedCodeFixSuggestedAction.FixAllFlavors.Actions.OfType<UnifiedFixAllCodeFixSuggestedAction>().Select(action => action.FixAllState.Scope.ToString());
-
-                    var title = string.Format(FeaturesResources.Fix_All_0, currentTitle);
-                    var command = new LSP.Command
-                    {
-                        CommandIdentifier = CodeActionsHandler.RunFixAllCodeActionCommandName,
-                        Title = title,
-                        Arguments = [new CodeActionResolveData(title, codeAction.CustomTags, request.Range, request.TextDocument, fixAllFlavors.ToArray())]
-                    };
-
-                    builder.Add(new LSP.CodeAction
+                    var codeActionPathList = codeActionPath.ToArray();
+                    var title = codeAction.Title;
+                    nestedCodeActions.Add(new LSP.CodeAction()
                     {
                         Title = title,
-                        Command = command,
                         Kind = codeActionKind,
                         Diagnostics = diagnosticsForFix,
-                        Data = new CodeActionResolveData(title, codeAction.CustomTags, request.Range, request.TextDocument, fixAllFlavors.ToArray())
+                        Data = new CodeActionResolveData(title, codeAction.CustomTags, request.Range, request.TextDocument, fixAllFlavors: null, nestedCodeActions: ImmutableArray<LSP.CodeAction>.Empty, codeActionPathList)
                     });
+
+                    if (suggestedAction is UnifiedCodeFixSuggestedAction unifiedCodeFixSuggestedAction && unifiedCodeFixSuggestedAction.FixAllFlavors is not null)
+                    {
+                        var fixAllFlavors = unifiedCodeFixSuggestedAction.FixAllFlavors.Actions.OfType<UnifiedFixAllCodeFixSuggestedAction>().Select(action => action.FixAllState.Scope.ToString());
+                        var fixAllTitle = string.Format(FeaturesResources.Fix_All_0, title);
+
+                        var command = new LSP.Command
+                        {
+                            CommandIdentifier = CodeActionsHandler.RunFixAllCodeActionCommandName,
+                            Title = fixAllTitle,
+                            Arguments = [new CodeActionResolveData(fixAllTitle, codeAction.CustomTags, request.Range, request.TextDocument, fixAllFlavors.ToArray(), nestedCodeActions: ImmutableArray<LSP.CodeAction>.Empty, codeActionPathList)]
+                        };
+
+                        nestedCodeActions.Add(new LSP.CodeAction()
```

Comment:
all this code action creation code looks pretty similar to the code starting at line 134 - could you extract and unify it all of that into a shared helper like `AddLspCodeAction` that 
1.  Creates the normal LSP action and adds it to the passed in list
2.  Creates the fix all LSP action and adds it to the passed in list